EditableGrid: remove dependence on QueryModel

labkey-alan opened the pull request:
#### Rationale
The EditableGrid depends directly on data from a QueryModel, which has added some unnecessary complexity. This PR puts all necessary data into the EditorModel, and refactors EditableGrid to reduce complexity.

#### Related Pull Requests
- https://github.com/LabKey/labkey-ui-components/pull/1534
- https://github.com/LabKey/labkey-ui-premium/pull/470
- https://github.com/LabKey/platform/pull/5687
- https://github.com/LabKey/limsModules/pull/469
- https://github.com/LabKey/testAutomation/pull/2016

#### Changes
- EditableGrid: remove unused or unnecessary props
  - allowFieldDisable
  - allowRemove
  - bordered
  - condensed
  - lockedRows
  - notDeletable
  - onSelectionChange
  - removeColumnTitle
  - showBulkTabOnLoad
  - striped
  - tabAdditionalBtn
  - tabBtnProps
- EditableGrid: remove dependence on QueryModel
- BulkAddUpdateForm: Improve types
- ExportMenu: Remove extraExportMenuOptions

labkey-alan committed:
Alert: Remove outdated comments

labkey-alan committed:
ExportMenu: Remove extraExportMenuOptions

This feature is no longer used

labkey-alan committed:
BulkAddUpdateForm: Improve types

- Remove defaults that were never overridden
- Add props to omitted list

labkey-alan committed:
- EditableGrid: remove unused props
  - allowFieldDisable
  - bordered
  - condensed
  - lockedRows
  - notDeletable
  - onSelectionChange
  - removeColumnTitle
  - showBulkTabOnLoad
  - striped
  - tabAdditionalBtn
  - tabBtnProps

labkey-alan committed:
EditorModel: rename columns to orderedColumns

labkey-alan committed:
Convert all usages of loadEditorModelData to EditableGridLoader

labkey-alan committed:
Remove loadEditorModelData from exports

labkey-alan committed:
EditorModel: remove isPasting

It is unused

labkey-alan committed:
Add init method to EditorModel

labkey-alan committed:
use FieldKey for genCellKey

labkey-alan committed:
expect FieldKey in parseCellKey

labkey-alan committed:
Simplify getReadOnlyRowCount

labkey-alan committed:
EditorModel: add getFolderValueForCellKey and getCellReadStatus

labkey-alan committed:
Move EditorModel.init to actions.ts::initEditorModel

labkey-alan committed:
EditableGrid: Fix issue with hideReadOnlyRows

labkey-alan committed:
Use columns from EditorModel

- Fix issue with getCellReadStatus

labkey-alan committed:
EditableGridPanelForUpdate: Fix prop types

labkey-alan committed:
Update release notes

labkey-alan committed:
EditableGrid: lint, remove some usages of data

labkey-alan committed:
BulkAddUpdateForm: improve default values

labkey-alan committed:
EditorModel: getRawDataFromGridData -> getDataForServerUpload
- The name was not accurate, it was never getting data from a grid model other than pk values
- Also deprecate getRawDataFromModel

labkey-alan committed:
EditorModel: Add isReadOnlyRow

Make getCellReadStatus more resilient

labkey-alan committed:
BulkAddUpdateForm: Remove data/dataKeys props

labkey-alan committed:
EditableGrid: don't use data/dataKeys when adding/selecting/pasting rows

labkey-alan committed:
EditableGrid/EditableGridChange: remove data/dataKeys props

labkey-alan committed:
Lint files

labkey-alan committed:
AssayImportPanels: remove dataModel

labkey-alan committed:
EditableGridPanelForUpdate: use initEditorModel

labkey-alan committed:
EntityInsertPanel: use initEditorModel

labkey-alan committed:
RunDataPanel: Fix issue with EditorModel initialization

labkey-alan committed:
getUpdatedDataFromEditableGrid: fix originalRows

labkey-alan committed:
AssayWizardModel: Fix issue with hasData

labkey-alan committed:
EditorModel: handle no pkCols

labkey-alan committed:
EditorModel.getCellReadStatus - make it more resilient

labkey-alan committed:
EditableGrid: always use columnMetdata from EditorModel

labkey-alan committed:
Remove unused methods and types

labkey-alan committed:
EditableGridPanel: Fix tabs, remove model prop
EditableGrid: remove queryInfo prop
EditorModel: remove id
BulkAddUpdateForm: remove queryInfo prop

labkey-alan committed:
EditableGrid: Remove columnMetadata prop

labkey-alan committed:
EditableGrid: Remove error prop

It was unused

labkey-alan committed:
EditableGrid: Remove insertColumns, updateColumns

EditableGridPanel: remove getUpdateColumns

labkey-alan committed:
initEditorModel: address issues with columns

labkey-alan committed:
EditableGrid: remove readOnlyColumns

labkey-alan committed:
EditableGridPanelForUpdate: add columnMetadata prop

labkey-alan committed:
EditorModel.getDataForServerUpload - remove unused args

labkey-alan committed:
EditableGrid: fix folder and row values passed to Cell

Cell: add columnMetadata prop, remove props that come from columnMetadata

labkey-alan committed:
Cell: more resilience around columnMetadata

labkey-alan committed:
EditableGrid: add TODO

labkey-nicka commented on packages/components/src/internal/components/assay/AssayWizardModel.ts:
```diff
@@ -218,23 +239,30 @@ export class AssayWizardModel
+        const rows = hasSamples ? selectedSamples : Map<string, Map<string, any>>({});
+        const loader = new AssayWizardModelEditableGridLoader(this.queryInfo, rows);
+        const dataModel = new QueryModel({ schemaQuery: this.queryInfo.schemaQuery }).mutate({ queryInfo: this.queryInfo });
+        return await initEditorModel(dataModel, loader, columnMetadata);
```
Great to finally have this usage refactored over.

labkey-nicka commented on packages/components/src/internal/components/editable/EditableGridPanelForUpdate.tsx:
```diff
@@ -144,6 +119,7 @@ export const EditableGridPanelForUpdate: FC<EditableGridPanelForUpdateProps> = p
         setIsSubmitting(true);
 
         try {
+            // TODO: I suspect we can skip passing originalRows since getDataForServerUpload appends folder/container
```
Comment marking TODO.

cnathe replied:
I agree that looking at the usage of this component in AssayGridPanel.tsx, the orignalRows here is just passed so that we can get the folder/container value for the rows. `getDataForServerUpload` does append that, but would `gridData.updatedRows` include it? It goes through `getUpdatedDataFromGrid` which will include values for the row that have been changed plus the id / pk field. Oh maybe the folder is included via the `altUpdateKeys`?

labkey-nicka commented on packages/components/src/internal/components/editable/models.ts:
```diff
@@ -126,45 +114,116 @@ export class EditorModel
+
+        if (pkFieldKey === undefined) return undefined;
+
+        return genCellKey(this.pkFieldKey, rowIndex);
```
nit: Pass `pkFieldKey` rather than recomputing `this.pkFieldKey`.

labkey-nicka commented on packages/components/src/internal/components/editable/models.ts:
```diff
@@ -126,45 +114,116 @@ export class EditorModel
+
+        if (cellKey === undefined) return undefined;
+
+        return this.cellValues.get(this.genPkCellKey(rowIndex))?.get(0).raw;
```
nit: Pass `cellKey` rather than recomputing `this.getPkCellKey(rowIndex)`.

labkey-nicka commented on packages/components/src/internal/components/editable/models.ts:
```diff
@@ -512,6 +543,20 @@ export class EditorModel
     }
 
+    isReadOnlyRow(rowIdx: number, readonlyRows: string[]): boolean {
+        const pkValue = this.getPkValue(rowIdx).toString();
```
Add defensive check for `this.getPkValue(rowIdx)?.toString()`. Also add `if (pkValue === undefined) return false;`.

labkey-nicka commented on packages/components/src/internal/components/editable/models.ts:
```diff
@@ -212,118 +272,102 @@ export class EditorModel
+    getRowValue(rowIdx: number, displayValues = true): Map<string, any> {
+        let row = Map<string, any>();
+
+        // TODO: NEED REVIEWER INPUT. Is it safe to use this.columnMap below? It includes pk columns (which we used to
```
I'd say let's go with it and let the callers figure out if things need to be stripped out.

labkey-nicka commented on packages/components/src/internal/components/editable/models.ts:
```diff
@@ -212,118 +272,102 @@ export class EditorModel
+        //  this.columnMap because insert cases won't have those columns, but update cases will, so things should just
+        //  work™.
+        this.columnMap.valueSeq().forEach(col => {
+            if (!col) return;
```
nit: I understand the defensiveness here but it seems like we should never allow undefined values to get into this map in the first place.

labkey-alan replied:
yeah, I didn't write that, it's an artifact of the original implementation, but I think you are right and it's safe to remove

labkey-nicka commented on packages/components/src/internal/components/editable/EditableGrid.tsx:
```diff
@@ -216,21 +221,20 @@ function inputCellFactory(
-        const className = classNames({ 'grid-col-with-width': hasCellWidthOverride(columnMetadata) });
-        const style = { textAlign: columnMetadata?.align ?? c.align ?? 'left' } as any;
-
+        // TODO: It would be really nice to drop the "row" prop in cell below, however the AssayTaskInputRenderer needs
```
I'm not so sure about this. There are cases where an individual cell may need to be able to see other values from the row. What would be the recourse if the cell didn't receive the data row?

labkey-alan replied:
The only case we have that needs the whole row is AssayTaskInputRenderer, it feels odd to have a prop that gets threaded through all of our Cell components, but is only needed by a single input renderer.

labkey-nicka commented on packages/components/src/internal/components/editable/actions.ts:
```diff
@@ -116,75 +96,97 @@ export const loadEditorModelData = async (
+    if (forUpdate) {
+        // If we're updating then we need to ensure that the pkCol is in the columnMap so things like readonlyRows
+        // will work.
+        const pkCol = queryInfo.getPkCols()[0];
```
QueryInfo isn't guaranteed to have a pkCol. Consider throwing if we do not resolve one in this case.

labkey-alan replied:
I've since changed this to add all pkCols and not just the first one. That way if we encounter a grid that has multiple it should work (though stuff like readOnlyRows would not be compatible without additional work).

labkey-nicka reviewed (commented):
Got most of the way through this review.

labkey-alan committed:
EditableGrid: Fix issue with isReadOnly

cnathe commented:
@labkey-alan  Manual testing issue found:
1. go to a sample type grid and select a sample with no status or available status and a sample with a locked status
2. edit in grid
3. make and update to just the available status
4. save / finish

This should be allowed since you aren't making any changes to the locked sample, but I get an error saying that the other sample is locked so it can't have its lineage updated (but I didn't update it)

labkey-alan committed:
initEditorModel: Add all pkCols and altUpdateKeys in forUpdate case

cnathe commented on packages/components/src/internal/components/editable/EditableGrid.tsx:
```diff
@@ -109,7 +109,8 @@ function computeSelectionCellKeys(
     for (let c = minColIdx; c <= maxColIdx; c++) {
         for (let r = minRowIdx; r <= maxRowIdx; r++) {
-            selectionCells.push(genCellKey(c, r));
+            const fieldKey = editorModel.columnMap.get(editorModel.orderedColumns.get(c)).fieldKey;
```
I see a few lines of code like this one. Thoughts on added an editorModel.getFieldKeyForColumnIndex helper?

cnathe replied:
Looks like there are about 8 usages of this in this file alone. Also, maybe a helper for the inverse might be useful as well to keep more of the usages of `orderedColumns` within the model itself (i.e. replace `orderedColumns.indexOf(fieldKey)` with `editorModel.getColumnIndexForFieldKey`)

labkey-alan replied:
If this PR gets merged with a lot of these usages, yes, I think it will make sense, but in general I want to get us away from using column indices. However, it may make sense to defer that work, and if I do, I'll add the helper.

labkey-alan replied:
I went ahead and added `getFieldKeyByIndex` and `getColumnByIndex` util methods to `EditorModel`

cnathe commented on packages/components/src/internal/components/editable/EditableGrid.tsx:
```diff
@@ -241,23 +245,19 @@ function inputCellFactory(
                     colIdx={colIdx}
-                    row={focused ? row : undefined}
+                    columnMetadata={columnMetadata}
+                    row={focused ? editorModel.getRowValue(rowIdx) : undefined}
```
when I first saw `getRowValue` I was assuming this was getting just a single value but it is really getting the full row object right? maybe rename to `getRowValues`?

labkey-alan replied:
We refer to the result of this method as "row" in all usages, and it roughly equates to getting a single row value from a QueryModel, so I think it makes sense to keep it as `getRowValue`.

cnathe commented on packages/components/src/internal/components/editable/EditableGrid.tsx:
```diff
@@ -621,13 +597,14 @@ export class EditableGrid extends PureComponent<EditableGridProps, EditableGridS
                         const maxInitialCell = parseCellKey(initialSelection[initialSelection.length - 1]);
-                        minColIdx = minInitialCell.colIdx;
-                        maxColIdx = maxInitialCell.colIdx;
+                        orderedColumns.indexOf(minInitialCell.fieldKey);
```
this line 600 seems out of place. I assume maybe just a copy-paste extra?

cnathe commented on packages/components/src/internal/components/editable/EditableGrid.tsx:
```diff
@@ -969,20 +942,20 @@ export class EditableGrid extends PureComponent<EditableGridProps, EditableGridS
                     index: GRID_EDIT_INDEX,
                     tableCell: true,
-                    title: this.props.removeColumnTitle,
+                    title: 'Delete',
```
Do we still have a use case where the editable grid has this "delete" column? We had a use case for some modal that configured some lookup values, but that one is gone. I can't recall a use case for this anymore.

labkey-alan replied:
I think you are correct and we no longer support this case, as I haven't been able to find any usages of the grid that have an inline delete row option. I'll remove this, and if we have a test fail as a result I'll revert my removal.

cnathe commented on packages/components/src/internal/components/editable/EditableGridPanel.tsx:
```diff
@@ -28,72 +42,34 @@ export interface EditableGridPanelProps extends SharedEditableGridPanelProps {
         [activeTab, onChange]
     );
-
-    // TODO: When EditableGridPanelDeprecated is removed we should be able to just pass model.rows and model.orderedRows
```
when was this TODO added? it was a long time coming, but happy to see this finally done!

labkey-alan replied:
I think it was added when I first added QueryModel support to the EditableGrid, so _years_ ago 😄

cnathe commented on packages/components/src/internal/components/editable/LookupCell.tsx:
```diff
@@ -161,18 +158,12 @@ export const LookupCell: FC<LookupCellProps> = memo(props => {
-    const containerPath_ =
-        containerPath ??
-        getValueFromRow(row?.toJS(), 'Folder')?.toString() ??
-        getValueFromRow(row?.toJS(), 'Container')?.toString();
```
I'm curious what change was made to make this not needed anymore? The use case here, I believe, was that if you selected rows from a grid where the rows were from different containers (app projects) and then did Edit in Grid, this could make sure that the QuerySelect used the rows containerPath for the query to get the select options instead of just using the current containerPath. Selenium tests were added for this case, so I guess we can see if those are passing.

labkey-alan replied:
Those selenium tests are passing. There are several editable grids (e.g. the DataClassEditableGridForUpdate) that add the Folder to the EditorModel columnMap if App.isProductProjectsEnabled is true. This code isn't needed because we grab the Folder from the EditorModel in the `inputCellFactory` if we're updating data _and_ there is a container value on the row.

cnathe commented on packages/components/src/internal/components/editable/actions.ts:
```diff
@@ -31,45 +21,33 @@ import {
+    rows: Record<string, any>,
+    columns?: QueryColumn[],
     forUpdate?: boolean
+    // TODO: When all non EditorModel.init usages of this are gone update this to only return cellValues and move it to
```
I am confused by this TODO as it looks like there is only one usage of this method now and it is within this same file

labkey-alan replied:
This was one of the first TODOs I added, and am still planning on moving this to the EditorModel, but want to get the other parts of this branch stabilized first.

cnathe commented on packages/components/src/internal/components/editable/actions.ts:
```diff
@@ -116,75 +94,107 @@ export const loadEditorModelData = async (
+    loader: EditableGridLoader,
+    columnMetadata?: Map<string, EditableColumnMetadata>
+): Promise<EditorModel> => {
+    // TODO: look at EditorModel.getColumns and make sure we're matching the behavior from there as well e.g. file
+    //  columns
+    const { columns: loaderColumns, queryInfo } = loader;
+    // TODO: Most EditableGridLoaders do not actually asynchronously fetch data (see note in EditableGridLoader), we
+    //  can most likely just drop loader as an arg, and have consumers pass in their QueryModel data (if any), and
+    //  skip having QueryModel as an arg.
+    const { data, dataIds } = await loader.fetch(queryModel);
+    // TODO: store rows and orderedRows on EditorModel so we can use it with getUpdatedDataFromEditableGrid
```
are any of these TODOs things you plan to look into for this PR?

labkey-alan replied:
Yes, basically all of them.

cnathe commented on packages/components/src/internal/components/editable/actions.ts:
```diff
@@ -116,75 +94,107 @@ export const loadEditorModelData = async (
-    const results = await Promise.all(
-        dataModels.map((dataModel, i) => initEditableGridModel(dataModels[i], editorModels[i], loaders[i], queryModel))
-    );
+        if (!hasContainerCol) {
```
why not just always include the containerCol (if it exists in the queryInfo as checked below)? It seems like putting it in the columnMap would be a good idea in all cases instead of relying on the columns to specifically request it.

labkey-alan replied:
That was what I was originally doing, however sometimes we have a container column on the columnMap and we explicitly want it to be visible, but readOnly, and not hidden, and those usages were modifying the QueryColumn object. So by always adding it we would be adding the version from the QueryInfo and not the one modified by the loader (or elsewhere). So, we only add it if it's not already in the columnMap so we don't accidentally add a different, unmodified, version. This caused a ton of test failures.

In general I think we shouldn't be touching any the columns on the QueryInfo anyway, it's one of the reasons why this branch has so many failures, because we have a ton of side effects. In an ideal world any overrides we want would be done via the EditableColumnMetadata, but that wasn't always an option so we have usages that don't use it, and some that use it while also editing the column objects. It's very messy.

cnathe commented on packages/components/src/internal/components/editable/actions.ts:
```diff
@@ -1450,68 +1264,38 @@ interface ParseLookupPayload {
 
     const byColumnValues = getPasteValuesByColumn(paste);
     const lookupColumnContainerCache = {};
+    // TODO: colMin may be wrong here because we're now storing hidden columns (e.g. RowId) in EditorModel
```
we are storing hidden columns in the EditorModel, but those aren't in the orderedColumns info right? So I would assume that the colMin, colMax, etc. here are based on the orderedColumns?

labkey-alan replied:
Yes, I believe this is totally safe, I was just being paranoid and writing these TODOs so that if I hit an test failure I could just look at all the places I left these notes to myself and hopefully track down the potential issues faster.

cnathe commented on packages/components/src/internal/components/editable/actions.ts:
```diff
@@ -1522,17 +1306,18 @@ async function insertPastedData(
+        let pkValue = getPkValue(row, editorModel.queryInfo);
+        if (!pkValue) pkValue = editorModel.getPkValue(rowIdx);
 
+        // TODO: given that we're storing all columns in EditorModel now this may not be correct.
```
as long as we are using orderedColumns below this should be correct I think

cnathe commented:
> @labkey-alan Manual testing issue found:
> 
> 1. go to a sample type grid and select a sample with no status or available status and a sample with a locked status
> 2. edit in grid
> 3. make and update to just the available status
> 4. save / finish
> 
> This should be allowed since you aren't making any changes to the locked sample, but I get an error saying that the other sample is locked so it can't have its lineage updated (but I didn't update it)

This has been resolved.

cnathe commented:
@labkey-alan Manual testing issue found:
1. from a sample type grid select some samples > Edit in Grid
EditableGridPanelForUpdateWithLineage.tsx:97 TypeError: updates.forEach is not a function
    at eval (models.ts:490:15)
    at eval (immutable.js:3015:46)
    at eval (immutable.js:1376:16)
    at ArrayMapNode.iterate.HashCollisionNode.iterate (immutable.js:1725:11)

This doesn't always repro. It seems to be based on the rows I select. The smallest set of selections that I can get to repro locally is having a locked aliquot selected which is from the current container and selecting another sample that is from a different container.

labkey-alan committed:
Revert changes to EditableGridLoaderFromSelection.tsx

labkey-alan commented:
> @labkey-alan Manual testing issue found:
> 
>     1. from a sample type grid select some samples > Edit in Grid
>        EditableGridPanelForUpdateWithLineage.tsx:97 TypeError: updates.forEach is not a function
>        at eval (models.ts:490:15)
>        at eval (immutable.js:3015:46)
>        at eval (immutable.js:1376:16)
>        at ArrayMapNode.iterate.HashCollisionNode.iterate (immutable.js:1725:11)
> 
> 
> This doesn't always repro. It seems to be based on the rows I select. The smallest set of selections that I can get to repro locally is having a locked aliquot selected which is from the current container and selecting another sample that is from a different container.

This is because somewhere in our products our types are wrong. I removed the code in our EditableGridLoaderFromSelection that coerced the incoming data to an Immutable `Map<string, any>`, because it was already typed as `Map<any ,any>` and the downstream consumers expected `Map<string, any>` so it felt safe. It turns out that it is not safe, and somewhere we're passing something like `Record<string,  any>` or similar, but Immutable clobbers types and it compiles anyway. I added a note to the EditableGridLoaderFromSelection documenting this issue.

labkey-alan committed:
PR feedback

labkey-alan committed:
Fix issue with addRowsPerPivotValue

labkey-alan committed:
Fix issue with Position Column in FM editable grids

labkey-alan committed:
EditorModel: force columnMetadata to use lowercase keys, add getColumnMetadata

labkey-alan committed:
Remove outdated TODOs, update other TODOs

labkey-alan committed:
EditableGrid: remove allowRemove

labkey-alan committed:
EditorModel: PR feedback

labkey-alan committed:
EditorModel: use lowercase fieldkeys, add getFieldKeyByIndex

labkey-alan committed:
Export genCellKey

genCellKey: lowercase fieldKey

labkey-alan committed:
EditableGridLoader: add extraColumns

labkey-alan committed:
Fix issues with addColumns, changeColumn

lowercase the fieldkeys!

labkey-alan committed:
Export parseCellKey

labkey-alan committed:
EditorModel.getColumnValues: lowercase fieldkey

labkey-alan committed:
PR feedback

labkey-nicka commented on packages/components/src/internal/components/editable/actions.ts:
```diff
@@ -3,22 +3,12 @@ import { fromJS, List, Map, OrderedMap, Set as ImmutableSet } from 'immutable';
-    parseCsvString,
-    parseScientificInt,
-} from '../../util/utils';
+import { caseInsensitive, isFloat, isInteger, parseCsvString, parseScientificInt } from '../../util/utils';
```
nit: Unused import on line 2 now due to these refactors.

labkey-alan committed:
Fix issues found by unit tests

labkey-alan committed:
ExportMenu.test.tsx - remove unnecessary test cases

labkey-alan committed:
QueryColumn: improve constructor types

labkey-alan committed:
Fix broken tests

labkey-nicka commented on packages/components/src/internal/components/editable/utils.test.ts:
```diff
@@ -975,20 +970,25 @@ describe('getValidatedEditableGridValue', () => {
     test('getSortedCellKeys', () => {
-        expect(sortCellKeys(['0-0', '1-1', '1-1', '0-1', '1-0'])).toStrictEqual(['0-0', '1-0', '0-1', '1-1']);
-        expect(sortCellKeys(['1-1', '1-15', '0-10', '1-5'])).toStrictEqual(['1-1', '1-5', '0-10', '1-15']);
+        // expect(sortCellKeys(['0-0', '1-1', '1-1', '0-1', '1-0'])).toStrictEqual(['0-0', '1-0', '0-1', '1-1']);
```
nit: Remove these comments now that the test is refactored.

labkey-nicka commented on packages/components/src/internal/components/editable/utils.test.ts:
```diff
@@ -10,7 +10,7 @@ import { makeTestQueryModel } from '../../../public/QueryModel/testUtils';
 import { QueryInfo } from '../../../public/QueryInfo';
 
-import { QueryColumn } from '../../../public/QueryColumn';
+import { QueryColumn, QueryLookup } from '../../../public/QueryColumn';
```
More of a question now that you've completed this initial pass. How do you feel the test coverage is in this file? Are there areas left uncovered? Are there new things introduced that we can circle back on to add coverage for here?

labkey-alan replied:
For utils I don't know that we need additional coverage. In theory we could add more test coverage for event handling like we have for addRows, addColumns, changeColumn, etc, however we do have test coverage for those events via our selenium tests so I am not sure how important it is to add coverage here.

We could also add more coverage for actions, like what we have for fillCell, however, like the utils stuff, we have a ton of selenium coverage, so I am not sure how much of a priority it really is.

Same would apply for the methods I added to the EditorModel, we could have some test coverage for them all, but again, we have a ton of selenium coverage for all of those methods.

labkey-alan committed:
Add test case for extraColumns

labkey-alan committed:
Prep for release

labkey-alan committed:
Prep for release
